aten/src/ATen/native/native_functions.yaml

```diff
@@ -3345,6 +3345,13 @@
   dispatch:
     DefaultBackend: celu_
 
+- func: siamese_loss(Tensor input1, Tensor input2, Tensor label, float margin=1.0, int reduction=Mean) -> Tensor
+  use_c10_dispatcher: fill
+  python_module: nn
+  dispatch:
+    CPU: siamese_loss
```

Comment:
I think the implementations for these functions are missing from this PR?

Reply:
I have added the functions in torch/nn/functional.py

Reply:
Quick clarification (I know there's a lot and it can be confusing) - the function in `torch/nn/functional.py` is the python implementation, and, in most cases, that just calls into a C++ implementation (alongside some size checks and `handle_torch_function()` logic like you've added). It looks like your python implementation currently calls:

```python
return torch.siamese_loss(input1, input2, label, margin)
```

The entry in `native_functions.yaml` you've added correctly declares a `siamese_loss()` function that python can call into, but there's a couple issues:
1. Since you specified `python_module: nn` above (which I agree with), the function will actually be available in python as `torch._C.nn.siamese_loss()`.
2. This function still needs C++ definitions underneath `aten/src/ATen/native`. The `CPU` and `CUDA` dispatch entries indicate which C++ functions define the implementation for each device.

So to be precise, for this `CPU: siamese_loss` line, I'd expect a C++ function named `siamese_loss()` that matches the interface you've specified above and implements the core logic, and it should likely be in `aten/src/ATen/native/Loss.cpp`.

Hopefully that all makes sense! Glad to help out more if needed :)

Reply:
Makes sense, will implement the c++ function that matches the interface. Will definitely reach out if there's anything